### DISCUSSION TITLE
macOS 12 (Monterey) compatibility

### DISCUSSION THREAD

**Discussion by adriweb** [CE-Programming/toolchain#344] [Announcements]
**UPDATE June 2022:**
Just download the [latest release](https://github.com/CE-Programming/toolchain/releases/latest), everything is ready to use :)

**UPDATE MARCH 2022:**
You can simply replace your local fasmg binary from the one in the upstream release at the bottom of https://flatassembler.net/download.php (**flat assembler g**). The binary will be in the source/macos/x64 folder.

---

The program `fasmg` (used as assembler and linker in the toolchain) is known not to work correctly anymore when run on macOS 12 (Monterey) due to a memory allocation issue. 

There is no fix yet however a workaround exists, which consists in injecting _GuardMalloc_ to fasmg when running it.

To do so automatically in the toolchain:
- Open the `meta/makefile.mk` with a text editor, located in your toolchain installation folder (for example inside `~/CEdev`)
- Replace the `FASMGLD` definition (line 83 currently) by `FASMGLD = DYLD_INSERT_LIBRARIES=/usr/lib/libgmalloc.dylib $(call NATIVEPATH,$(BIN)/fasmg)`

This has only been tested on Intel macs so far, not Apple Silicon (M1 etc) yet.

**Comment by Meorge** [comment]
This workaround didn't work for me.
My lines 80-85 of `meta/makefile.mk`:
```
QUOTE_ARG = "$(subst ",',$1)"#'
else
NATIVEPATH = $(subst \,/,$1)
FASMGLD = DYLD_INSERT_LIBRARIES=/usr/lib/libgmalloc.dylib $(call NATIVEPATH,$(BIN)/fasmg)
CONVBIN = $(call NATIVEPATH,$(BIN)/convbin)
CONVIMG = $(call NATIVEPATH,$(BIN)/convimg)
```
Running `make` for the `hello_world` example code yields the following output:
```
➜  hello_world make
[compiling] src/main.c
[convimg] icon.png
[linking] bin/DEMO.bin
Error: not enough memory to complete the assembly.
make: *** [bin/DEMO.bin] Error 3
```

I'm on an Intel iMac running macOS Monterey 12.2.1, using v9.2.2 of the toolchain (latest as of this post).

**Reply by adriweb** [reply]
We'll have to check if the lib in question is present, maybe it's not there by default (without Xcode?). Although that would mean `make` wouldn't have worked... so I'm not sure what's going on.

Ping me on IRC/Discord when you're available so we can investigate all that a bit more.

**Reply by Meorge** [reply]
For others who find this thread: we determined that, at least on my computer, [disabling SIP](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection) along with the instructions provided in the original post allowed it to complete successfully.

**Reply by adriweb** [reply]
Well, SIP is enabled for me and everything works just fine, so it must be something else... but so far we don't know what.

```
18:37:15 [:~] % csrutil status
System Integrity Protection status: enabled.
```

Doesn't hurt (well...) to try with SIP disabled if someone else runs into the same issue though.

That said, upstream fasmg may have an actual fix for the issue: https://board.flatassembler.net/topic.php?t=22003

**Comment by adriweb** [comment]
An actual solution was [found upstream](https://board.flatassembler.net/topic.php?p=221428#221428), and I've tested it successfully on my macOS setup :D

I suppose we'll be able to have it for the next toolchain release.
